### ISSUE TITLE
Double send headers and cookies

### ISSUE THREAD

**Issue by oleg578** [phalcon/cphalcon#3193]
Method Phalcon\Mvc\Application handle send headers and cookies
https://github.com/phalcon/cphalcon/blob/2.0.0/phalcon/mvc/application.zep#L453
and return Phalcon\HTTP\Response
After that when we send send Response, headers and cookies are send again...
https://github.com/phalcon/cphalcon/blob/2.0.0/phalcon/http/response.zep#L595

If I wrote in application normal code

``` php
$application->main()->handle()->send();
```

I get double headers and cookies

```
curl -i http://localhost:8000/admin
HTTP/1.1 200 OK
Host: localhost:8000
Connection: close
X-Powered-By: PHP/5.6.4
Set-Cookie: PHPSESSID=8avt3lufu8rc2e72ne5stvjdmvdtta1k; path=/
Expires: Thu, 19 Nov 1981 08:52:00 GMT
Cache-Control: no-store, no-cache, must-revalidate, post-check=0, pre-check=0
Pragma: no-cache
Set-Cookie: __urs=pg%2BgYpjdpSZlQBgVazybtIzGtYWyX2MGTXLcL88%2BLENfRCUBwBW%2Buk0%3D; path=/; httponly
Set-Cookie: __urs=EHKDQ8t0oWjawsKzXYGbSBVfq2DS46lMMUroKvt2TcyXIGKKHNkPQUM%3D; path=/; httponly
Content-type: text/html; charset=UTF-8
```

I found this in 
Phalcon 1.3.4 and 2.0.0
##

**Comment by olivier-monaco** [comment]
I don't see any double header in your "curl output".

Sending headers is done by Phalcon\Http\Response\Headers::send() with check if headers are already sent. So it's not possible.

The same with Phalcon\Http\Response\Cookies::send().

Here is my "curl output":
<pre>
$ curl -i http://localhost/
HTTP/1.1 200 OK
Date: Thu, 08 Jan 2015 21:59:21 GMT
Server: Apache/2.2.22 (Debian)
Vary: Accept-Encoding
Content-Length: 2
Content-Type: text/html; charset=UTF-8

OK
</pre>

Do you have a use case?

**Comment by oleg578** [comment]
Curl don't show double headers, but show double cookies.

This behavior we can see in source code (see my issue).

Why need Response->send , if headers and cookies sending in Application->hendle ?
It is logically, when Application->handle return Response object, but  why it sends headers and cookies? 
It should not be. I don't uderstood this behaviaor.
Only Response should be send all to browser.
Or Application->handle should be send headers, cookies and content...
The issue can see in the first example og doc
```php
echo $application->handle()->getContent(); 
```

**Comment by olivier-monaco** [comment]
There are 2 parts in your issues

1) double headers/cookies

I really think it's not. You can use telnet to see the raw data.

<pre>
$ telnet localhost 80
GET / HTTP/1.1
Host: localhost

HTTP/1.1 200 OK
Date: Thu, 08 Jan 2015 23:39:40 GMT
Server: Apache/2.2.22 (Debian)
Vary: Accept-Encoding
Content-Length: 2
Connection: close
Content-Type: text/html; charset=UTF-8

OK
</pre>

As I told you, the code avoid double headers.

Your double cookies can come from another problem. Can you dump the content of $response->getCookies() and check what the _cookies property contents?

2) The way Application works.

I'm agree with you, Application::handle must send all or nothing (nothing is better). Maybe @phalcon  can give us more info. But, again, calling send on the returned response will not send headers/cookies twice.

**Comment by oleg578** [comment]
See, please, links to source in my first post.
Headers and cookies are sending in Application->handle and in Response->send.
I loss 2 work day out of  this behavior

**Comment by olivier-monaco** [comment]
I've already seen your links and also more code. And like I told you, there's protection to avoid double headers and double cookies:
https://github.com/phalcon/cphalcon/blob/2.0.0/phalcon/http/response/headers.zep#L90
https://github.com/phalcon/cphalcon/blob/2.0.0/phalcon/http/response/cookies.zep#L258

But... after testing, there's a problem, as you told me.

This comes from headers_sent() that will return true only when at least one byte of the HTTP content has been sent... But when Application::handle() returns no content is send. So headers_sent() returns false.

So, it's a bug!

**Comment by oleg578** [comment]
I think enough delete strings in Application->handle where headers and cookies sending.

**Comment by olivier-monaco** [comment]
Not so easy... We need to remove:
<pre>
		/**
		 * Headers and Cookies are automatically send
		 */
		response->sendHeaders();
		response->sendCookies();
</pre>

But also to move:
<pre>
		/**
		 * Calling beforeSendResponse
		 */
		if typeof eventsManager == "object" {
			eventsManager->fire("application:beforeSendResponse", this, response);
		}
</pre>

into Reponse::send(). But Response is not EventsAwareInterface... Working on this with #3249.

**Comment by olivier-monaco** [comment]
But the event name is linked to application. I can't move it to the Response class.

So I can let it at the end of Application::handle (but we do not guarantee to send the response) or we call the send method on the response at the end of this method in place of sendHeaders and sendCookies.

What is best?

**Comment by oleg578** [comment]
My opinion is  -
1. Conponents Application and Response can be isolated.
This satisfies the classical structure   Request -> Application -> Respponse.

Then Application can return only Response object without any events of Response.
And main code string will be as 
```php
$application->main()->handle()->send();

```
instead of
```php
echo $application->main()->handle()->getContent();
```
Because now with this string we reject the work flow paradigm -
Receive request -> Handle data with Application -> build Response -> send Response.
All manipulations with response can be in Response class.
Of course, all events of response can be in Response class.
Would be nice if the Response class the EventsAwareInterface interface.
I'm so sorry. I do not have in-depth knowledge of CPhalcon.
But I hope my thoughts are clear.
Would be nice if this problem be solved in 2.0
Now I got work API (JSON) controller 
```php
public function onConstruct()
    {
        $this->view->disable();
        $this->response->setContentType('application/json', 'utf-8');
    }

public function afterExecuteRoute()
    {
        $this->response->setJsonContent($this->content, JSON_NUMERIC_CHECK);
        $this->response->send();
    }
...
```
But this isn't like for me.

**Comment by dzubchik** [comment]
@oleg578 , I have recently faced with this issue. So, I agree with you, that app should return request object only, but not send headers and cookies.
